src/Spago/Command/Fetch.purs

```diff
@@ -442,29 +442,48 @@ toAllDependencies = foldl (Map.unionWith (\l _ -> l)) Map.empty
-  Git.fetchRepo package tempDir >>= case _ of
-    Left err -> die err
-    Right _ -> do
-      logDebug $ "Repo cloned. Moving to " <> localPackageLocation
-      FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
-      FS.moveSync { src: tempDir, dst: localPackageLocation }
-
-      -- Note: the package might have been cloned with a tag, but we stick the commit hash in the lockfiles
-      -- so we need to make a copy to a location that has the commit hash too.
-      -- So we run getRef here and then do a copy if the ref is different than the original one
-      -- (since it might be a commit to start with)
-      logDebug $ "Checking if we need to copy the package to a commit hash location..."
-      Git.getRef (Just localPackageLocation) >>= case _ of
-        Left err -> die err
-        Right ref -> do
-          when (ref /= package.ref) do
-            let commitHashLocation = Config.getPackageLocation name (GitPackage $ package { ref = ref })
-            logDebug $ "Copying the repo also to " <> commitHashLocation
-            FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
-            FS.copyTree { src: localPackageLocation, dst: commitHashLocation }
+  logDebug $ "Copying repo to " <> localPackageLocation
+  FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
+  FS.copyTree { src: repoCacheLocation, dst: localPackageLocation }
+
+  -- Note: the package might have been cloned with a tag, but we stick the commit hash in the lockfiles
+  -- so we need to make a copy to a location that has the commit hash too.
+  -- So we run getRef here and then do a copy if the ref is different than the original one
+  -- (since it might be a commit to start with)
+  logDebug $ "Checking if we need to copy the package to a commit hash location..."
+  commitHash <- Git.getRef (Just localPackageLocation) >>= rightOrDie
+  when (commitHash /= package.ref) do
+    let commitHashLocation = Config.getPackageLocation name (GitPackage $ package { ref = commitHash })
+    logDebug $ "Copying the repo also to " <> commitHashLocation
+    FS.copyTree { src: localPackageLocation, dst: commitHashLocation }
+  where
+  repoCacheLocation = Path.concat [ Paths.localCacheGitPath, Config.fileSystemCharEscape package.git ]
+
+  ensureRepoCloned = unlessM (FS.exists repoCacheLocation) do
+    tempDir <- mkTemp' (Just $ printJson Config.gitPackageCodec package)
+    logDebug $ "Cloning repo in " <> tempDir
+    Git.fetchRepo package tempDir >>= rightOrDie_
+
+    logDebug $ "Repo cloned. Moving to " <> repoCacheLocation
+    FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
+    FS.moveSync { src: tempDir, dst: repoCacheLocation }
+
+  ensureRefCheckedOut = do
+    logDebug $ "Checking out " <> package.ref
+    { offline } <- ask
+    Git.checkout { repo: repoCacheLocation, ref: package.ref } >>= case _, offline of
```

Comment:
Doing a checkout on the cached location means that we can't execute two operations on the same repo in parallel (I think in principle that can happen. Even if it doesn't happen in the same Spago process, it can happen if you run two Spago commands at the same time).

I feel like we should checkout the repo only once it's copied in the final location, but that means that we wouldn't know when to pull.
So we could either:
- always pull
- or find out if the ref exists in the cached location (without performing a checkout - I am not sure `git` has a command for that), and pull if it doesn't

Reply:
> I am not sure git has a command for that

Ah apparently that's `git cat-file -t <ref>`

Reply:
Oh, I didn't realize you wanted to support parallel work on the same repo. Sure, will do. 

To find out if the ref is there, we can also use `git show`

Reply:
Ok, finally I'm done.